tendermint/src/lite/verifier.rs

```diff
@@ -284,42 +246,100 @@ pub fn verify_header<TS, SH, C, L, S, R>(
 ) -> Result<(), Error>
 where
     TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
     L: TrustThreshold,
-    S: Store<TrustedState = TS>,
     R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
-    C: Commit,
+    S: Store<TrustedState = TS>,
+{
+    // Fetch the latest state and ensure it hasn't expired.
+    // Note we only check for expiry once in this
+    // verify_and_update_bisection function since we assume the
+    // time is passed in and we don't access a clock internall.
+    // Thus the trust_period must be long enough to incorporate the
+    // expected time to complete this function.
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    // TODO: consider fetching the header we're trying to sync to and
+    // checking that it's time is less then `now + X` for some small X.
+    // If not, it means that either our local clock is really slow
+    // or the blockchains BFT time is really wrong.
+    // In either case, we should probably raise an error.
+    // Note this would be stronger than checking that the untrusted
+    // header is within the trusting period, as it could still diverge
+    // significantly from `now`.
+
+    // inner recursive function which assumes
+    // trusting_period check is already done.
+    _verify_and_update_bisection(untrusted_height, trust_threshold, req, store)
+}
+
+// inner recursive function for verify_and_update_bisection.
+// see that function's docs.
+fn _verify_and_update_bisection<TS, SH, C, L, R, S>(
```

Comment:
This `_` is not a good sign. If this is internal then the two intents can be split and made into distinct functions. Unless I miss a point why this needs to be in one.

Reply:
So just `internal_update_bisection`?

Reply:
If the rename above goes through this can just be `verify_and_update_bisection`.